### PR TITLE
CLN: allow label memviews to be None

### PR DESCRIPTION
**Pull request by mzeitlin11**
Saves some unnecessary memory allocations, plus moving towards making `rank_sorted_1d` easier to call. Also have done something similar for allowing `mask` to be None, but split that off this pr to keep it simpler.

### CODE REVIEW

File: pandas/_libs/algos.pyx [hunk 1]
```diff
@@ -451,11 +448,8 @@ def nancorr_spearman(ndarray[float64_t, ndim=2] mat, Py_ssize_t minp=1) -> ndarr
                             with gil:
                                 # We need to slice back to nobs because rank_1d will
                                 # require arrays of nobs length
-                                labels_nobs = np.zeros(nobs, dtype=np.int64)
-                                rankedx = rank_1d(np.array(maskedx)[:nobs],
-                                                  labels=labels_nobs)
-                                rankedy = rank_1d(np.array(maskedy)[:nobs],
-                                                  labels=labels_nobs)
+                                rankedx = rank_1d(np.array(maskedx)[:nobs])
+                                rankedy = rank_1d(np.array(maskedy)[:nobs])
```
Comment (jbrockmendel) [inline pandas/_libs/algos.pyx:452]: not sure if it makes a difference, but np.array -> np.asarray?

File: pandas/_libs/algos.pyx [hunk 2]
```diff
@@ -1096,17 +1090,18 @@ def rank_1d(
 cdef void rank_sorted_1d(
     float64_t[::1] out,
     int64_t[::1] grp_sizes,
-    const intp_t[:] labels,
     const intp_t[:] sort_indexer,
     # Can make const with cython3 (https://github.com/cython/cython/issues/3222)
     rank_t[:] masked_vals,
     const uint8_t[:] mask,
-    TiebreakEnumType tiebreak,
```
Comment (jbrockmendel) [inline pandas/_libs/algos.pyx:1104]: as long as you're shifting these around, can you make everything from here-ish on keyword-only

**Comment by jbrockmendel** [comment]
Small comments, generally looks very nice

File: pandas/_libs/algos.pyx (hunk 1 above)
Comment (mzeitlin11) [reply pandas/_libs/algos.pyx]: good catch, was an oversight no reason to copy here

File: pandas/_libs/algos.pyx (hunk 2 above)
Comment (mzeitlin11) [reply pandas/_libs/algos.pyx:1104]: I don't think cdef functions support keyword-only enforced by "*" (unless you have a workaround, I just tried and saw a cython compile error :)

Changed callers to treat it as keyword-only from after mask, though, which should help

**Comment by jreback** [comment]
thanks @mzeitlin11

**Pull request merged**